UnityProject/Assets/Scripts/Weapons/Gun.cs

```diff
@@ -0,0 +1,889 @@
+		/// </summary>
+		[Tooltip("The amount of times per second this weapon can fire")]
+		public double FireRate;
+
+		/// <summary>
+		/// If suicide shooting should be prevented (for when user inadvertently drags over themselves during a burst)
+		/// </summary>
+		[Tooltip("If suicide shots should be prevented (if the user accidentally mouses over themselves during a shot")]
+		protected bool AllowSuicide;
+
+		//TODO connect these with the actual shooting of a projectile
+		/// <summary>
+		/// The max recoil angle this weapon can reach with sustained fire
+		/// </summary>
+		public float MaxRecoilVariance;
+
+		/// <summary>
+		/// The projectile fired from this weapon
+		/// </summary>
+		[Tooltip("The projectile fired from this weapon")]
+		public GameObject Projectile;
+
+		//TODO: make this dependent on the mag/projectile
+		/// <summary>
+		/// The amount of projectiles spawned per shot
+		/// </summary>
+		[SerializeField]
+		private int ProjectilesFired = 1;
+
+		//TODO: make this dependent on the mag used/projectile fired
+		/// <summary>
+		/// The traveling speed for this weapons projectile
+		/// </summary>
+		[Tooltip("The speed of the projectile")]
+		public int ProjectileVelocity;
+
+		/// <summary>
+		/// Describes the recoil behavior of the camera when this gun is fired
+		/// </summary>
+		[Tooltip("Describes the recoil behavior of the camera when this gun is fired")]
+		public CameraRecoilConfig CameraRecoilConfig;
+
+		/// <summary>
+		/// The firemode this weapon will use (burst,semi,auto)
+		/// </summary>
+		[Tooltip("The firemode this weapon will use")]
+		public WeaponType WeaponType;
+
+		/// <summary>
+		/// The sound this gun makes when shooting
+		/// </summary>
+		[SerializeField]
+		private AddressableAudioSource firingSound;
+
+		/// <summary>
+		/// The sound this gun makes when out of ammo
+		/// </summary>
+		[SerializeField]
+		private AddressableAudioSource outOfAmmoSound;
+
+		/// <summary>
+		/// The sound this gun makes when out of empty
+		/// </summary>
+		[SerializeField]
+		private AddressableAudioSource emptySound;
+
+		/// <summary>
+		/// Used only in server, the queued up shots that need to be performed when the weapon FireCountDown hits
+		/// 0.
+		/// </summary>
+		private Queue<QueuedShot> queuedShots;
+
+		/// <summary>
+		/// We don't want to eject the magazine or reload as soon as the client says its time to do those things - if the client is
+		/// firing a burst, they might shoot the whole magazine then eject it and reload before server is done processing the last shot.
+		/// Instead, these vars are set when the client says to eject or load a new magazine, and the server will only process
+		/// the actual unload / load (updating MagNetID) once the shot queue is empty.
+		/// </summary>
+		private bool queuedUnload = false;
+		private uint queuedLoadMagNetID = NetId.Invalid;
+
+		private RegisterTile registerTile;
+		private ItemStorage itemStorage;
```

Comment:
Codacy found an issue: [Remove the field 'itemStorage' and declare it as a local variable in the relevant methods.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6280689)